### ISSUE TITLE
Host/Exec targets cannot load shared libs on remote

### ISSUE THREAD

**Issue by werkt** [bazelbuild/bazel#10809]
### Description of the problem / feature request:

Using a host or exec configured binary with shared lib dependencies on remote fails at runtime

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

Use a `cc_binary` with `linkstatic = 0` that links against a `cc_library` as a `tools` or `exec_tools` member of a `genrule.` See attached workspace archive.

The host/exec bin path relative to the exec root used to execute the binary, whether in a `genrule` or the `executable` of a `ctx.actions.run`, is in a location which cannot resolve a remote-input specified exec root `_solib_<cpu>` which contains the shared lib. The action is shipped with an input root that does not contain the exec-root wide `_solib_<cpu>`.

The *only* way that the non-remote sandbox would be able to execute these targets is if this path is available, representing a differentiation between the action composition and the sandbox inputs (i.e. the sandbox shows things that the remote is not told about).

The remote action also contains a .runfiles for the target, which also contains the binary itself under its root, and can be used successfully to execute the command in place of the specified binary and locate the runfiles-relative copy of the `_solib_<cpu>` for its needed libraries.

### What operating system are you running Bazel on?

Ubuntu 18.04

### What's the output of `bazel info release`?

2.1.0

### Any other information, logs, or outputs that you want to share?

[This is an archive](https://github.com/bazelbuild/bazel/files/4216618/bazel-remote-runfiles.tar.gz) of a workspace which demonstrates this problem. Run locally, in standalone or sandbox, `bazel build :baz` completes successfully. Run remotely, it fails.

I can teach my remote system to recognize the runfiles pattern and replace it as a first executable with the runfiles relative target, and even teach it to mutate the genrule pattern, but this seems like a huge workaround for not specifying an input tree or runfiles-relative command correctly.

**Comment by ulfjack** [comment]
If I remember correctly, there was an issue with the standard C++ linker which typically puts absolute paths to the .so into the binary and that obviously causes problems. Again, if I remember correctly, Google created an upstream C++ linker/compiler patch a few years ago. I'm not sure whether that was before 18.04, whether it's available on 18.04, and whether the Bazel C++ toolchain supports it. I *think* setting LD_LIBRARY_PATH such that all .so in the uploaded tree in the remote execution system might serve as a workaround.

Maybe someone from RBE can comment on this. @ola-rozenfeld, do you someone know who might know more?

**Comment by werkt** [comment]
Just a clarification, the *default* behavior for gcc/clang+binutils (with or without gold) has been to pull the SONAME from a linked .so, or the basename if it does not contain one, and place that into the NEEDED list of a binary (no absolute paths there, I've never seen any pathing whatsoever in NEEDED lists of any kind). The default link RUNPATH/RPATH is unpopulated, and bazel provides only a package-relative population of the `runtime_library_search_directories` for injection into these. The default (and our) C++ toolchain implementation from bazel injects a `$ORIGIN/<directory>` for each of those directories.

I believe that the link specification of runtime behavior here is correct, and that there is no absolute specification of paths, but the clear difference between either the input sets or the behavior to local sandbox and remote execution for the binary's invocation is troubling. It only works because the sandbox is provided additional inputs that remote is not, or sandboxing is providing its own special input tree completely on its own.

**Comment by ulfjack** [comment]
Could it be that the runtime linker is following the symlinks into the exec root and then looking up files relative to that? Could that explain why it works with sandboxing, but not remote execution?

**Comment by emfree** [comment]
Hi there, has there been any progress on the issue? Or alternatively do the C++ rules experts have any pointers on where to start for a fix patch in bazel? This is a pretty confusing bug when using remote execution.

**Comment by oquenchil** [comment]
Hi Eben,

If I understand correctly what is going on here, this is an issue with the rpath.

That doesn't happen for cc_test because of this [feature](https://cs.opensource.google/bazel/bazel/+/master:tools/cpp/unix_cc_toolchain_config.bzl;drc=8b9f74649512ee17ac52815468bf3d7e5e71c9fa;l=448).

That is controlled by is_cc_test. I have to find out from my colleague why the build variable was added for that feature and why it wasn't simply enabled for cc_test in the rule implementation.

If I am able to safely remove the build variable we can do so and the fix for cc_binary would be:
A) People that need this would enable the feature themselves with the `features` attribute
B) If linkstatic=0, the feature should be enabled automatically from inside the rule

I would go with B). But if you want to send a fix patch, I will let you know what the best course of action is by next week.

**Comment by oquenchil** [comment]
So the only reason that I see why the build variable exists is because it should be mutually exclusive to use EXEC_ORIGIN or ORIGIN. Perhaps the fastest fix for now is for the cc_binary with linkstatic=0 to also set the is_cc_test variable, after renaming it to something else.

Did you check if changing ORIGIN to EXEC_ORIGIN solved your issue?

**Comment by benjaminp** [comment]
`EXEC_ORIGIN` is useless outside of Google because the patch never made it into upstream glibc. (https://sourceware.org/legacy-ml/libc-alpha/2017-10/msg01109.html is the last attempt I'm aware of to do so.)

**Comment by oquenchil** [comment]
Thank you Benjamin.

**Comment by carlosgalvezp** [comment]
@werkt 
> bazel provides only a package-relative population of the runtime_library_search_directories for injection into these
Could you ellaborate on how `runtime_library_search_directories` is populated? I want to use a hermetic GCC toolchain instead the one installed in the host, and I can't find a way to populate `runtime_library_search_directories` with the correct paths.

**Comment by fmeum** [comment]
@sgowroji This is fixed in Bazel 6.0.0:

```
❯ USE_BAZEL_VERSION=5.3.2 bazel build //... --spawn_strategy=remote --remote_executor=grpc://localhost:8080
INFO: Invocation ID: b3cbe4b0-18d2-470f-a13f-210a666ecdd5
INFO: Analyzed 3 targets (38 packages loaded, 305 targets configured).
INFO: Found 3 targets...
ERROR: /home/fhenneke/Downloads/bazel-remote-runfiles/BUILD:14:8: Executing genrule //:baz failed: (Exit 127): bash failed: error executing command /bin/bash -c 'source external/bazel_tools/tools/genrule/genrule-setup.sh; bazel-out/host/bin/foo > bazel-out/k8-fastbuild/bin/baz.out'
bazel-out/host/bin/foo: error while loading shared libraries: liblibbar.so: cannot open shared object file: No such file or directory
INFO: Elapsed time: 3.950s, Critical Path: 1.40s
INFO: 22 processes: 13 internal, 9 remote.
FAILED: Build did NOT complete successfully

❯ USE_BAZEL_VERSION=6.0.0 bazel build //... --spawn_strategy=remote --remote_executor=grpc://localhost:8080
Starting local Bazel server and connecting to it...
INFO: Invocation ID: 26564303-54d0-48af-bf12-674bae6fdcb3
INFO: Analyzed 3 targets (37 packages loaded, 294 targets configured).
INFO: Found 3 targets...
INFO: Elapsed time: 5.678s, Critical Path: 0.81s
INFO: 22 processes: 12 internal, 10 remote.
INFO: Build completed successfully, 22 total actions
```

**Comment by sgowroji** [comment]
@fmeum Thanks for adding the pointers. @werkt We are closing this issue now. Please verify it with the latest 6.0.0 release and  Feel free to reach us if you still have any specific queries on the same.
